### ISSUE TITLE
Letsencrypt renewal issue since Virtualmin 7.20.0

### ISSUE THREAD

**Issue by websmurf** [virtualmin/virtualmin-gpl#864]
Most likely related to this change: 
> Add ability to check the resolvability of alternative names before issuing a Let's Encrypt certificate

I have a setup with a lot of sites that have their root record to a certain ip-address, and also aliasses for the www and mail subdomains. For example: bandhosting.be

The certificate setup in virtualmin is: 
<img width="1426" alt="image" src="https://github.com/user-attachments/assets/f975d88d-1397-426f-9ac6-a1f3532b5960">

The result of a certificate renewal is, that the mail.bandhosting.be and www.bandhosting.be domains will not be requested due to the cleanup:
<img width="869" alt="image" src="https://github.com/user-attachments/assets/f8a9b3b0-fb2b-45c6-83f8-f832dd5c6884">

They all do point to the same ip though: 
<img width="1192" alt="image" src="https://github.com/user-attachments/assets/7a4d0087-1fc6-49c7-8b16-c77051cd6181">
<img width="1195" alt="image" src="https://github.com/user-attachments/assets/014a1772-4cef-4c2d-a194-eace64efe2d9">
<img width="1208" alt="image" src="https://github.com/user-attachments/assets/608b3c54-99b9-47f2-9db1-5421d444cc93">

The result is a certificate, that is only valid for the root domain, and not the www and mail subdomains: https://bandhosting.be/
https://www.bandhosting.be/
https://mail.bandhosting.be/

**Comment by iliajie** [comment]
Hello Adam!

Thanks for the heads up!

Adam, this is indeed a bug on the Virtualmin side, as when we use `check_external_dns()` API, we only consider IPv4 address and not CNAME record! To work around this issue for now, you should turn off the resolvability check.

Jamie, to be clear, we need to consider CNAME record as legitimate as well, e.g.:

```
dig www.bandhosting.be @8.8.8.8

; <<>> DiG 9.10.6 <<>> www.bandhosting.be @8.8.8.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 28793
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.bandhosting.be.		IN	A

;; ANSWER SECTION:
www.bandhosting.be.	3600	IN	CNAME	bandhosting.be.
bandhosting.be.		3600	IN	A	37.97.169.184

;; Query time: 113 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Sun Jul 21 19:47:43 EEST 2024
;; MSG SIZE  rcvd: 77
```

**Comment by iliajie** [comment]
Adam, please check this https://github.com/virtualmin/virtualmin-gpl/commit/a287e6c3a4519c9e7051d145cf7bbb141ceef8f9 patch and see if it solves the problem for you?

**Comment by jcameron** [comment]
Thanks Ilia for the fix!

**Comment by iliajie** [comment]
Welcome!

**Comment by websmurf** [comment]
Just applied the patch and after a webmin restart: 

<img width="833" alt="image" src="https://github.com/user-attachments/assets/6716c132-3b24-425a-9e55-14ce22093a5b">

Thanks for the swift response!

**Comment by iliajie** [comment]
I'm glad it worked!

**Comment by iliajie** [comment]
> Sorry to barge in but i think this should be chnaged, the function is not future proof and will cause issues in the near future.

This is optional and can be disabled upon request in Virtualmin.

> it use external tools and not Net::DNS, output could be unreliable

We use `dig` with external DNS to check if the world (i.e., Let's Encrypt) can see the records. There is nothing unreliable about that.

> it does not support ipv6

Well, IPv6-only configurations are still pretty rare. However, I agree that we need to consider it as well.

> it use hardcoded google DNS 8.8.8.8, this will fail on secure network that force certain DNS server with firewall rules.

Alright, I would make the external DNS configurable as well.

**Comment by iliajie** [comment]
> > it does not support ipv6
> 
> Well, IPv6-only configurations are still pretty rare. However, I agree that we need to consider it as well.

Jamie, please check this https://github.com/virtualmin/virtualmin-gpl/commit/6bc2355d560cbc1c9a821fd64585db542007c1d1 patch and this https://github.com/virtualmin/virtualmin-gpl/commit/1e65c67f06c83d0a286308811bcc2a77f2b533d6 patch too.

**Comment by aqueos** [comment]
thanks, i am not directly concerned by those but the cname hit me on several domains.

external tools can change their output that is why i dont like using them in a program.

i saw that we can disable it but as the option is on by default even on allready configured domains i would have to check all domains of all sites on all servers to see if a cname is there (or ipv6 ) and go manualy change it one by one. :(

thanks for the quick answer.

**Comment by iliajie** [comment]
I'd like to point out that DNS resolution check is a fallback in case normal request failed.

**Comment by iliajie** [comment]
> are you sure it test dns only if the letsencrypt renew fails ?

I have checked the code, and to my surprise, we have different logic when Let's Encrypt is initially requested (upon domain creation) versus during renewal. For renewals, we indeed always test DNS resolvability!

This is fixed in this https://github.com/virtualmin/virtualmin-gpl/commit/26256ff8625259bf5e91b6dfced19d42fafc7e8d patch.

**Comment by jcameron** [comment]
Does this only effect renewals for the default domain?

**Comment by iliajie** [comment]
All renewals.

**Comment by jcameron** [comment]
@iliajie the current code will use the `letsencrypt_nodnscheck` field to skip the DNS check, which can be configured in the UI. It's not set by default which causes the DNS check to be done if the `dig` command is installed, even on renewal. Your patch would actually change the current behavior and not do checks on renewal, which I don't think is what we want assuming that it can be done reliability.

@aqueos I actually prefer to use external commands if possible, as relying on installed Perl modules just increases the installation complexity and dependencies of Webmin

**Comment by aqueos** [comment]
i agree that this check is a good thing to prevent issues with letsencrypt. The problem is the code that is not finished and has been deployed in production without support for  things like cname or ipv6 so it locks the renewal of all letsencrypt using those :(

@jcameron ok for dig , i would not, but this is a matter of personal preferences :)

thanks to all for the quick action, i think after the test it should be pushed in production as it will block a lot of people that use the basic dns configuration that is:

domain.com IN A x.X.X.X 
+
www.domain.com IN CNAME domain.com. 

and each day can lead to sites being down with no ssl :)

**Comment by iliajie** [comment]
> @iliajie the current code will use the `letsencrypt_nodnscheck` field to skip the DNS check, which can be configured in the UI.

Yes, that's what I thought as well.

> It's not set by default which causes the DNS check to be done if the `dig` command is installed, even on renewal.

Yes, but that is something I decided to change. In my current view, I don't think we should re-check records in renewals that were previously successful.

> Your patch would actually change the current behavior and not do checks on renewal, which I don't think is what we want assuming that it can be done reliability.

Oh, alright! But why would we want to have it rechecked in renewals as well, assuming it worked in the past?

> The problem is the code that is not finished and has been deployed in production without support for things like cname or ipv6 so it locks the renewal of all letsencrypt using those 

Well, it is fixed now!

>  i think after the test it should be pushed in production as it will block a lot of people that use the basic dns configuration that is:

Yes, I agree! We will try to make a new Virtualmin 7.20.1 release as soon as we agree on everything being discussed.

**Comment by digitaldutch** [comment]
I ran in the same issue. The auto certificate update failed.

```
Checking hostnames for resolvability ..
Fatal Error!
Failed to request certificate : None of the hostnames could be resolved! 

virtual-server/letsencrypt.cgi (line 111)
```

The domain was named test.somedomain.com Set with a cname. The work around was switching off the checking.

**Comment by iliajie** [comment]
> The domain was named test.somedomain.com Set with a cname. The work around was switching off the checking.

Sorry about that! We will fix it in the immediate Virtualmin 7.20.2 release!

**Comment by PitWenkin** [comment]
> > The domain was named test.somedomain.com Set with a cname. The work around was switching off the checking.
> 
> Sorry about that! We will fix it in the immediate Virtualmin 7.20.2 release!

Please make sure to update repositories as soon as possible… cf https://github.com/virtualmin/virtualmin-gpl/issues/855

**Comment by PitWenkin** [comment]
> > > The domain was named test.somedomain.com Set with a cname. The work around was switching off the checking.
> > 
> > 
> > Sorry about that! We will fix it in the immediate Virtualmin 7.20.2 release!
> 
> Please make sure to update repositories as soon as possible… cf #855

7.20.2 fixes the problem **and** is available via repositories 👍

**Comment by aqueos** [comment]
Thanks for your work.

  Just one last thing that i recall now. When you buy a domain in France a lot of time it configure a basic zone with ip4 and ipv6 pointer to a welcome page of the registrar. 

  This can lead to failure on letsencrypt on ipv4 host as letsencrypt prioritize ipv6 and fail. Got a lot of that on Ovh.

  So my point is : if you want to add foolproof check you can add a check that if you got a ipv6 on the domain then the host should have one too because, if not, this will fail. 

  I guess the reverse is true but i never saw a domain with ipv4 and no ipv6 on a ipv6 host :)

best regards.
Ghislain

**Comment by jcameron** [comment]
We do have that check already actually - if the domain has an IPv6 record in DNS but it's not enabled in Apache, Virtualmin won't even attempt to request the cert.

**Comment by jcameron** [comment]
> Have you progressed on it?

The check you requested is already in place in Virtualmin
